File: tinychain/vms/brainfuck.py
from collections import defaultdict
import sys

def is_valid_bf(code):
    language = '><+-.,[]'

    # Validate program.
    # Verify code only contains alphabet from the language.
    for char in code:
        if char not in language:
            return False
    return True

def create_jump_table(chars):
    jump_table = {}
    left_positions = []

    position = 0
    for char in chars:
        if char == '[':
            left_positions.append(position)

        elif char == ']':
            left = left_positions.pop()
            right = position
            jump_table[left] = right
            jump_table[right] = left
        position += 1

    return jump_table

def dump_mem(mem):
    for key, value in mem.items():
        print("{:04d}: {}".format(key, value))
    
    # buf = ""
    # MEM_STEP = 8
    # for i in range((len(mem) // MEM_STEP) + 1):
    #     buf += "{:04d} ".format(i)
    #     for j in range(MEM_STEP):
    #         buf += chr(mem[i+j])
    # return buf

def brainfuck_fmt(code):
    output_ = ""
    depth = 0
    indent = "  "
    for char in code:
        if char == '[':
            output_ += "\n"
            output_ += (depth * indent) + "["
            depth += 1
            output_ += "\n"
            output_ += depth * indent
        elif char == ']':
            output_ += "\n"
            depth -= 1
            output_ += (depth * indent) + "]"
            output_ += "\n"
            output_ += depth * indent
        else:
            output_ += char
    return output_

# The out of gas exception.
class OutOfGasException(Exception):
    pass

# Runs a Brainfuck interpreter on code.
# `memory` is all zeros by default.
def brainfuck_run(code, input_="", mem=defaultdict(int), gas_limit=250000000000, debugger=False):
    # data pointer
    dp = 0

    # output buffer
    output = ""
    # input buffer
    input_i = 0

    # debug log
    class DebugStream:
        def __init__(self):
            self.log = ""
        
        def write(self, x):
            self.log += x
    
    debug_stream = DebugStream()
    def debug(x):
        debug_stream.write(x)
        if False:
            sys.stdout.write(x)

    # instruction pointer
    ip = 0
    # jump register
    jmp_register = -1
    # jump table
    jmp_table = create_jump_table(code)

    # gas metering.
    GAS_USE_TABLE = {
        'memory': 3,
        'compute': 1
    }
    gas_used = 0

    while True:
        if len(code) - 1 < ip:
            # end of program.
            break
        
        if gas_limit < gas_used:
            raise OutOfGasException("gas_limit = {}, gas_used = {}".format(gas_limit, gas_used))

        opcode = code[ip]
        
        gas_used += GAS_USE_TABLE["compute"]
        
        # "00100 +"
        debug("{:05d} {}\n".format(ip, opcode))

        if debugger:
            input("op={} dp={:05d} d={:02d} >".format(opcode, dp, mem[dp]))

        if opcode == '>':
            # Increment the data pointer by one
            dp += 1
        elif opcode == '<':
            # Decrement the data pointer by one
            dp -= 1
        elif opcode == '+':
            # Increment the byte at the data pointer by one. 
            gas_used += GAS_USE_TABLE["memory"]
            mem[dp] += 1
        elif opcode == '-':
            # Decrement the byte at the data pointer by one. 
            gas_used += GAS_USE_TABLE["memory"]
            mem[dp] -= 1
        elif opcode == '.':
            # Output the byte at the data pointer. 
            # print(mem[dp])
            output += chr(mem[dp] % 256)
        elif opcode == ',':
            # Accept one byte of input, storing its value in the byte at the data pointer. 
            if input_i < len(input_) - 1:
                mem[dp] = ord(input_[input_i])
                input_i += 1
            else:
                # TODO. exception?
                mem[dp] = 0
        elif opcode == '[':
            # Jump If Zero
            # Jumps to the matching ] instruction if the value of the current cell is zero
            if mem[dp] == 0:
                jmp_register = jmp_table[ip]
        elif opcode == ']':
            # Jump If Not Zero
            # Jumps to the matching [ instruction if the value of the current cell is nonzero
            if mem[dp] != 0:
                jmp_register = jmp_table[ip]
        elif opcode == ';':
            # Comment.
            # Skip all characters until the next semicolon.
            while code[ip] != ';':
                ip += 1

        # Advance instruction pointer if not a jump instruction.
        if jmp_register == -1:
            ip += 1
        else:
            debug("{:05d} JUMP {}\n".format(ip, jmp_register))
            ip = jmp_register
            jmp_register = -1
        
        continue
    
    return (output, debug_stream.log, mem, gas_used)

class BrainfuckVM:
    def __init__(self):
        self.memory = defaultdict(int)
    
    def eval(self, code="", gas_limit=0):
        memory1 = self.memory.copy()
        (output, debug, mem, gas_cost) = brainfuck_run(code, memory1, gas_limit)

    def apply(self, code="", gas_limit=0):
        memory1 = self.memory.copy()
        (output, debug, mem, gas_cost) = brainfuck_run(code, "", mem=memory1, gas_limit=gas_limit)
        self.memory = memory1
        return (output, gas_cost)
    
    def dump_memory(self):
        buf = ""
        for key, value in self.memory.items():
            buf += "{:04d}: {}\n".format(key, chr(value))
        return buf


if __name__ == "__main__":
    # print "Hello world" to screen.
    # (output, debug, mem, gas_cost) = brainfuck_run("++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.")
    # (output, debug, mem, gas_cost) = brainfuck_run(">>>>>++")
    
    
    
    # (output, debug, mem, gas_cost) = brainfuck_run(
    #     open('brainfuck/bf-interpreter.bf').read(),
    #     # input=open('brainfuck/helloworld.bf').read()
    #     input_="++"
    # )


    (output, debug, mem, gas_cost) = brainfuck_run(
        open('brainfuck/solong.bf').read(),
        # "+++++>++++++temp0[-]x[temp0+x-]y[x+y-]temp0[y+temp0-]",
        # input=open('brainfuck/helloworld.bf').read()
        # input_="++"
    )


    # (output, debug, mem) = brainfuck_run("[->+<]")
    # print(brainfuck_fmt("+[-->-[>>+>-----<<]<--<---]>-.>>>+.>>..+++[.>]<<<<.+++.------.<<-.>>>>+."))
    # (output, debug, mem, gas_cost) = brainfuck_run("+[-->-[>>+>-----<<]<--<---]>-.>>>+.>>..+++[.>]<<<<.+++.------.<<-.>>>>+.")

    print("output: {}".format(output))
    print("gas_cost: {}".format(gas_cost))
    print(dump_mem(mem))
    # print(mem)


    

File: tinychain-go/core/vm/brainfuck.go
package vm

import (
	"fmt"
)

var language = "><+-.,[]"

func isValidBF(code string) bool {
	for _, char := range code {
		if !contains(language, char) {
			return false
		}
	}
	return true
}

func contains(s string, c rune) bool {
	for _, char := range s {
		if char == c {
			return true
		}
	}
	return false
}

func createJumpTable(chars string) map[int]int {
	jumpTable := make(map[int]int)
	leftPositions := []int{}

	for pos, char := range chars {
		switch char {
		case '[':
			leftPositions = append(leftPositions, pos)
		case ']':
			left := leftPositions[len(leftPositions)-1]
			leftPositions = leftPositions[:len(leftPositions)-1]
			jumpTable[left] = pos
			jumpTable[pos] = left
		}
	}
	return jumpTable
}

func dumpMem(mem map[int]int) {
	for key, value := range mem {
		fmt.Printf("%04d: %c\n", key, value)
	}
}

func brainfuckFmt(code string) string {
	output := ""
	depth := 0
	indent := "  "
	for _, char := range code {
		switch char {
		case '[':
			output += "\n" + string(repeat(' ', depth*len(indent))) + "[\n"
			depth++
			output += string(repeat(' ', depth*len(indent)))
		case ']':
			output += "\n"
			depth--
			output += string(repeat(' ', depth*len(indent))) + "]\n"
			output += string(repeat(' ', depth*len(indent)))
		default:
			output += string(char)
		}
	}
	return output
}

func repeat(char rune, count int) []rune {
	repeated := make([]rune, count)
	for i := 0; i < count; i++ {
		repeated[i] = char
	}
	return repeated
}

type OutOfGasException struct {
	Message string
}

func (e *OutOfGasException) Error() string {
	return e.Message
}

func brainfuckRun(code string, input string, mem map[int]int, gasLimit int64, debugger bool) (string, string, map[int]int, int64, error) {
	dp := 0
	output := ""
	inputIdx := 0
	ip := 0
	jmpRegister := -1
	jmpTable := createJumpTable(code)

	GAS_USE_TABLE := map[string]int64{
		"memory":  3,
		"compute": 1,
	}

	gasUsed := int64(0)

	for ip < len(code) {
		if gasLimit < gasUsed {
			return "", "", mem, gasUsed, &OutOfGasException{Message: fmt.Sprintf("gas_limit = %d, gas_used = %d", gasLimit, gasUsed)}
		}

		opcode := rune(code[ip])
		gasUsed += GAS_USE_TABLE["compute"]

		if debugger {
			fmt.Printf("op=%c dp=%05d d=%02d >\n", opcode, dp, mem[dp])
			fmt.Scanln()
		}

		switch opcode {
		case '>':
			dp++
		case '<':
			dp--
		case '+':
			gasUsed += GAS_USE_TABLE["memory"]
			mem[dp]++
		case '-':
			gasUsed += GAS_USE_TABLE["memory"]
			mem[dp]--
		case '.':
			output += string(rune(mem[dp] % 256))
		case ',':
			if inputIdx < len(input) {
				mem[dp] = int(input[inputIdx])
				inputIdx++
			} else {
				mem[dp] = 0
			}
		case '[':
			if mem[dp] == 0 {
				jmpRegister = jmpTable[ip]
			}
		case ']':
			if mem[dp] != 0 {
				jmpRegister = jmpTable[ip]
			}
		case ';':
			for code[ip] != ';' {
				ip++
			}
		}

		if jmpRegister == -1 {
			ip++
		} else {
			ip = jmpRegister
			jmpRegister = -1
		}
	}

	return output, "", mem, gasUsed, nil
}

type BrainfuckVM struct {
	memory map[int]int
}

func NewBrainfuckVM() *BrainfuckVM {
	return &BrainfuckVM{memory: make(map[int]int)}
}

func (vm *BrainfuckVM) Eval(code string, gasLimit int64) (string, int64, error) {
	memCopy := copyMemory(vm.memory)
	output, _, _, gasCost, err := brainfuckRun(code, "", memCopy, gasLimit, false)
	return output, gasCost, err
}

func (vm *BrainfuckVM) Apply(code string, gasLimit int64) (string, int64, error) {
	memCopy := copyMemory(vm.memory)
	output, _, mem, gasCost, err := brainfuckRun(code, "", memCopy, gasLimit, false)
	if err == nil {
		vm.memory = mem
	}
	return output, gasCost, err
}

func (vm *BrainfuckVM) DumpMemory() string {
	buf := ""
	for key, value := range vm.memory {
		buf += fmt.Sprintf("%04d: %c\n", key, value)
	}
	return buf
}

func copyMemory(mem map[int]int) map[int]int {
	memCopy := make(map[int]int)
	for k, v := range mem {
		memCopy[k] = v
	}
	return memCopy
}
